File: csharp/Palindromo.cs
using System;
using System.Linq;

class Palindromo
{
    static string Normalizar(string texto)
    {
        return new string(texto.ToLower().Where(char.IsLetterOrDigit).ToArray());
    }

    static bool EhPalindromo(string texto)
    {
        string normalizado = Normalizar(texto);
        char[] invertidoArray = normalizado.ToCharArray();
        Array.Reverse(invertidoArray);
        string invertido = new string(invertidoArray);
        return normalizado == invertido;
    }

    static void Main(string[] args)
    {
        Console.OutputEncoding = new System.Text.UTF8Encoding(false);

        if (args.Length != 1)
        {
            Console.WriteLine("Uso: Palindromo.exe \"texto a verificar\"");
            Environment.Exit(1);
        }

        string texto = args[0];
        bool resultado = EhPalindromo(texto);

        Console.WriteLine($"Texto original:  {texto}");
        Console.WriteLine($"Texto comparado: {Normalizar(texto)}");
        Console.WriteLine(resultado ? "É palíndromo!" : "Não é palíndromo.");

        Environment.Exit(resultado ? 0 : 1);
    }
}

File: lua/palindromo.lua
local function normalizar(texto)
    texto = texto:lower()
    local normalizado = texto:gsub("[^%a%d]", "")
    return normalizado
end

local function inverter(texto)
    return texto:reverse()
end

local function ehPalindromo(texto)
    local normalizado = normalizar(texto)
    return normalizado == inverter(normalizado)
end

local function main(arg)
    if #arg ~= 1 then
        print('Uso: lua5.4 palindromo.lua "texto a verificar"')
        os.exit(1)
    end

    local texto = arg[1]
    local resultado = ehPalindromo(texto)

    print("Texto original:  " .. texto)
    print("Texto comparado: " .. normalizar(texto))
    print(resultado and "É palíndromo!" or "Não é palíndromo.")

    os.exit(resultado and 0 or 1)
end

main(arg)
